domain/src/main/java/org/oppia/android/domain/question/QuestionAssessmentCalculation.kt

```diff
@@ -0,0 +1,99 @@
+package org.oppia.android.domain.question
+
+import org.oppia.android.app.model.FractionGrade
+import org.oppia.android.app.model.UserAssessmentPerformance
+import javax.inject.Inject
+
+/**
+ * Private class that computes the state of the user's performance at the end of a practice session.
+ * This class is not thread-safe, so it is the calling code's responsibility to synchronize access
+ * to this class.
+ */
+internal class QuestionAssessmentCalculation private constructor(
+  private val viewHintPenalty: Int,
+  private val wrongAnswerPenalty: Int,
+  private val maxScorePerQuestion: Int,
+  private val internalScoreMultiplyFactor: Int,
+  private val questionSessionMetrics: List<QuestionSessionMetrics>,
+  private val skillIdList: List<String>
+) {
+  /** Compute the overall score as well as the score and mastery per skill. */
+  internal fun computeAll(): UserAssessmentPerformance {
+    return UserAssessmentPerformance.newBuilder().apply {
+      totalFractionScore = computeTotalScore()
+      putAllFractionScorePerSkillMapping(computeScoresPerSkill())
+      // TODO(#3067): Set up finalMasteryPerSkillMapping
+    }.build()
+  }
+
+  private fun computeTotalScore(): FractionGrade =
+    FractionGrade.newBuilder().apply {
+      val allQuestionScores: List<Int> = questionSessionMetrics.map { it.computeQuestionScore() }
+      pointsReceived = allQuestionScores.sum().toDouble() / internalScoreMultiplyFactor
+      totalPointsAvailable =
+        allQuestionScores.size.toDouble() * maxScorePerQuestion / internalScoreMultiplyFactor
```

Comment:
Do we need to divide both the numerator and denominator of the FractionGrade by the factor? Wont the fraction stay equivalent regardless?

Reply:
@Vinita @BenHenning My reasoning for dividing both the numerator and denominator by the multiplier was to make sure the final FractionGrade's numerator and denominator is consistent with the values that we would have gotten on Oppia web. Since Chantel said she may want to display the score as a fraction (e.g. "you got x / y points"), I thought it would be better to do that here so the app layer doesn't have to simplify it. 

I made a separate constant for this multiplier because it's possible that the maxScorePerQuestionConstant can be different from the multiplier if the scoring system were to the change in the future. As an example right now, for mastery calculations, maxMasteryGainPerQuestion is 10, but we multiplied all the constants by a factor of 100 to maintain integer values. In the case that they are different (like for masteries), we would need to use both values to calculate the final scores/masteries. I thought keeping this multiplier separate would be good for consistency and also leave room for future changes so the math still checks out. 

Let me know what you think, thanks!